embassy-mcxa/src/trng.rs

```diff
@@ -0,0 +1,620 @@
+
+        regs().scmisc().write(|w| unsafe {
+            w.lrun_max()
+                .bits(config.long_run_limit_max)
+                .rty_ct()
+                .bits(config.retry_count)
+        });
+
+        regs()
+            .mctl()
+            .modify(|_, w| w.dis_slf_tst().variant(config.self_test.into()));
+
+        regs().sdctl().write(|w| unsafe {
+            w.samp_size()
+                .bits(config.sample_size)
+                .ent_dly()
+                .bits(config.entropy_delay)
+        });
+
+        regs()
+            .osc2_ctl()
+            .modify(|_, w| w.trng_ent_ctl().variant(config.osc_mode.into()));
+
+        regs().mctl().modify(|_, w| w.prgm().disable());
+
+        let _ = regs().ent(7).read().bits();
+    }
+
+    fn start() {
+        regs().mctl().modify(|_, w| w.trng_acc().set_bit());
+    }
+
+    fn stop() {
+        regs().mctl().modify(|_, w| w.trng_acc().clear_bit());
+    }
+
+    fn blocking_wait_for_generation() {
+        while regs().mctl().read().ent_val().bit_is_clear() {}
+    }
+
+    fn fill_chunk(chunk: &mut [u8]) {
+        let mut entropy = [0u32; 8];
+
+        for (i, item) in entropy.iter_mut().enumerate() {
+            *item = regs().ent(i).read().bits();
+        }
+
+        let entropy: [u8; 32] = unsafe { core::mem::transmute(entropy) };
+
+        chunk.copy_from_slice(&entropy[..chunk.len()]);
+    }
+
+    // Blocking API
+
+    /// Fill the buffer with random bytes, blocking version.
+    pub fn blocking_fill_bytes(&mut self, buf: &mut [u8]) {
+        if buf.is_empty() {
+            return; // nothing to fill
+        }
+
+        Self::start();
+        for chunk in buf.chunks_mut(32) {
+            Self::blocking_wait_for_generation();
+            Self::fill_chunk(chunk);
+        }
+        Self::stop();
+    }
+
+    /// Return a random u32, blocking version.
+    pub fn blocking_next_u32(&mut self) -> u32 {
+        Self::start();
+        Self::blocking_wait_for_generation();
+        let result = regs().ent(0).read().bits();
+
+        // New random bytes are generated only after reading ENT7
+        let _ = regs().ent(7).read().bits();
+        Self::stop();
+
+        result
+    }
+
+    /// Return a random u64, blocking version.
+    pub fn blocking_next_u64(&mut self) -> u64 {
+        Self::start();
+        Self::blocking_wait_for_generation();
+
+        let mut result = u64::from(regs().ent(0).read().bits()) << 32;
+        result |= u64::from(regs().ent(1).read().bits());
+
+        // New random bytes are generated only after reading ENT7
+        let _ = regs().ent(7).read().bits();
+        Self::stop();
+
+        result
+    }
+}
+
+impl Drop for Trng<'_> {
+    fn drop(&mut self) {
+        // reset the TRNG
+        regs().mctl().write(|w| w.rst_def().set_bit());
+    }
+}
+
+/// TRNG Async Driver
+pub struct AsyncTrng<'d> {
+    _peri: Peri<'d, TRNG0>,
+}
+
+impl<'d> AsyncTrng<'d> {
```

Comment:
Any reason you went for `AsyncTrng` instead of `Trng<'d, M: Mode>`? It's a little inconsistent with other peripherals.

Reply:
I can certainly change :) In fact, I had the mode typestate before. No real reason to remove it, honestly.

Reply:
Let me know if you'd prefer the `Mode` typestate.